core-api/src/main/java/com/optimizely/ab/config/audience/UserAttribute.java

```diff
@@ -98,27 +94,26 @@ public Boolean evaluate(ProjectConfig config,
         } catch(UnknownValueTypeException e) {
             if (!attributes.containsKey(name)) {
                 //Missing attribute value
-                String message = reasons.addInfo("Audience condition \"%s\" evaluated to UNKNOWN because no value was passed for user attribute \"%s\"", this, name);
-                logger.debug(message);
+                logger.debug("Audience condition \"{}\" evaluated to UNKNOWN because no value was passed for user attribute \"{}\"", this, name);
             } else {
                 //if attribute value is not valid
                 if (userAttributeValue != null) {
-                    String message = reasons.addInfo("Audience condition \"%s\" evaluated to UNKNOWN because a value of type \"%s\" was passed for user attribute \"%s\"",
+                    logger.warn(
+                        "Audience condition \"{}\" evaluated to UNKNOWN because a value of type \"{}\" was passed for user attribute \"{}\"",
                         this,
                         userAttributeValue.getClass().getCanonicalName(),
                         name);
-                    logger.warn(message);
                 } else {
-                    String message = reasons.addInfo("Audience condition \"%s\" evaluated to UNKNOWN because a null value was passed for user attribute \"%s\"", this, name);
-                    logger.debug(message);
+                    logger.debug(
+                        "Audience condition \"{}\" evaluated to UNKNOWN because a null value was passed for user attribute \"{}\"",
+                        this,
+                        name);
                 }
             }
         } catch (UnknownMatchTypeException | UnexpectedValueTypeException e) {
-            String message = reasons.addInfo("Audience condition \"%s\" " + e.getMessage(), this);
-            logger.warn(message);
+            logger.warn("Audience condition \"{}\" " + e.getMessage(), this);
         } catch (NullPointerException e) {
-            String message = reasons.addInfo("attribute or value null for match %s", match != null ? match : "legacy condition");
-            logger.error(message, e);
+            logger.error("attribute or value null for match {}", match != null ? match : "legacy condition", e);
```

Comment:
if the only reason to have reasons is for debugging, then, not having these doesn't make sense.  also, we need to make sure we are doing the same for other sdks.  We were including attribute failures before.

Reply:
I agree it's a good to have in reasons, so was added before. Found that this is the only meaningful message from conditions. It's quite expensive to apply the unused reasons to all Conditions, that's why I decided to drop this from reasons. We provide other relevant reasons like "does not meet audience condition", ... and they can take a look at log messages for investigation if needed.